### PR TITLE
Use template / generic type for annotating registries object types.

### PR DESCRIPTION
**Pull request by ADmad**
https://psalm.dev/docs/annotating_code/templated_annotations/
https://arnaud.le-blanc.net/post/phpstan-generics.html

Support for generic types in phpstan is still WIP and partially available using it's `master` branch currently.

Edit: phan supports it too https://github.com/phan/phan/wiki/Generic-Types

### CODE REVIEW

**Comment by markstory** [comment]
Looks good outside of the test failures.

File: src/Core/ObjectRegistry.php [hunk 1]
```diff
@@ -180,10 +181,10 @@ protected function _checkDuplicate(string $name, array $config): void
      * This method should construct and do any other initialization logic
      * required.
      *
-     * @param string|object $class The class to build.
+     * @param string|TObject $class The class to build.
      * @param string $alias The alias of the object.
      * @param array $config The Configuration settings for construction
-     * @return object
+     * @return TObject
```
Comment (dereuromark) [inline src/Core/ObjectRegistry.php]: Do IDEs also understand this?

File: src/Log/LogEngineRegistry.php [hunk 2]
```diff
@@ -84,9 +86,7 @@ protected function _create($class, string $alias, array $settings): LoggerInterf
             return $instance;
         }
 
-        throw new RuntimeException(
-            'Loggers must implement Psr\Log\LoggerInterface.'
-        );
+        throw new RuntimeException('Loggers must instanceof ' . LoggerInterface::class);
```
Comment (dereuromark) [inline src/Log/LogEngineRegistry.php]: implement?

Comment (ADmad) [reply src/Log/LogEngineRegistry.php]: Fixed

File: src/Core/ObjectRegistry.php (hunk 1 above)
Comment (ADmad) [reply src/Core/ObjectRegistry.php]: Dunno. People can submit issue for their IDE of choice if it doesn't.

Comment (ADmad) [reply src/Core/ObjectRegistry.php]: Dunno. People can submit issue for their IDE of choice if it doesn't.

**Pull request merged**